Changes required to compile with TBB 2021.2

gartung opened the pull request:
#### PR description:
Changes required to compile with TBB v2021.2
Requires cms-sw/cmsdist#6792
#### PR validation:
Tested with TBB v2021.2

gartung committed:
Initial changes to compile with TBB 2021.2

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22179

- This PR adds an extra 68KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22179/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22179/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22179/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22179/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

gartung committed:
code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22180

- This PR adds an extra 68KB to repository

cmsbuild commented:
A new Pull Request was created by  @gartung (Patrick Gartung) for master.

It involves the following packages:

CondCore/Utilities
DetectorDescription/Core
FWCore/PluginManager
FWCore/Services
FWCore/Utilities
HLTrigger/Timer
PhysicsTools/TensorFlow


@perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @riga, @wddgit, @Martin-Grunewald, @mmusich, @fabiocos this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

gartung committed:
Add cstring to define strstr

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22194

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

gartung committed:
For header check

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22229

- This PR adds an extra 80KB to repository

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

makortel commented on PhysicsTools/TensorFlow/interface/TBBThreadPool.h:
```diff
@@ -27,7 +27,7 @@ namespace tensorflow {
 
     explicit TBBThreadPool(int nThreads = -1)
-        : nThreads_(nThreads > 0 ? nThreads : tbb::task_scheduler_init::default_num_threads()), numScheduleCalled_(0) {
+        : nThreads_(nThreads > 0 ? nThreads : tbb::global_control::max_allowed_parallelism), numScheduleCalled_(0) {
```
I suspect this is not correct. The `tbb::global_control::max_allowed_parallelism` is an enumeration
```cpp
class global_control {
public:
    enum parameter {
        max_allowed_parallelism,
```
https://github.com/oneapi-src/oneTBB/blob/9e15720bc7744f85dff611d34d65e9099e077da4/include/oneapi/tbb/global_control.h#L57-L60
and the code here should use the actual number of threads (that is communicated for TensorFlow via the `NumThreads()` function).

On the other hand, I suspect the `default_num_threads()` to be potentially incorrect as well (I'd naively expect the `NumThreads()` needing to return the number of threads that can be used, not the number of logical cores on the machine).

makortel replied:
Here is some past discussion
https://github.com/cms-sw/cmssw/pull/28711#issuecomment-579233555
https://github.com/cms-sw/cmssw/pull/28711#issuecomment-579283411
https://github.com/cms-sw/cmssw/pull/28711#issuecomment-579288424

This PR should not get too involved in "what is correct here" though.

gartung replied:
How about `tbb::global_control::active_value(tbb::global_control::max_allowed_parallelism)`
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/FWCore/Services/plugins/InitRootHandlers.cc#L854

makortel replied:
@dan131riley pointed to
https://github.com/cms-sw/cmssw/blob/b4bb14f6573e803ecbe516d306d651d9774d48b7/FWCore/Services/plugins/InitRootHandlers.cc#L852-L854
I think we should do the same here (and leave any further tuning for later).

gartung commented on FWCore/Framework/test/dependentrecord_t.cppunit.cc:
```diff
@@ -38,6 +38,7 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include <cstring>
```
yes

makortel replied:
Now this line can be removed, right?

gartung commented on FWCore/Utilities/interface/hash_combine.h:
```diff
@@ -38,6 +38,15 @@ namespace edm {
+    std::size_t operator()(const std::pair<const unsigned int, const unsigned int>& a) const noexcept {
+      return edm::hash_value(a.first, a.second);
+    }
+  };
```
OK

gartung commented on FWCore/Utilities/interface/zero_allocator.h:
```diff
@@ -0,0 +1,55 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
```
OK

gartung committed:
move StdPairHasher to its own header
add comments on origin or zero allocator
use tbb::global_control::active_value() to get number of threads

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22249

ERROR: Build errors found during clang-tidy run.
```
 ;
--
FWCore/Utilities/interface/std_pair_hasher.h:10:8: error: redefinition of 'StdPairHasher' [clang-diagnostic-error]
struct StdPairHasher {
       ^
FWCore/Utilities/interface/hash_combine.h:42:10: note: previous definition is here
--
FWCore/Utilities/interface/std_pair_hasher.h:19:2: error: expected ';' after struct [clang-diagnostic-error]
}
 ^
 ;
--
CondCore/Utilities/bin/conddb_test_gt_perf.cpp:508:23: error: no matching constructor for initialization of 'tbb::global_control' [clang-diagnostic-error]
  tbb::global_control init(tbb::global_control::active_value(tbb::global_control::max_allowed_parallelism), nThrF);
                      ^
/cvmfs/cms-ib.cern.ch/nweek-02677/slc7_amd64_gcc900/external/tbb/2020_U3-ljfedo/include/tbb/global_control.h:33:5: note: candidate constructor not viable: no known conversion from 'std::size_t' (aka 'unsigned long') to 'tbb::interface9::global_control::parameter' for 1st argument
--
CondCore/Utilities/bin/conddb_test_gt_perf.cpp:571:23: error: no matching constructor for initialization of 'tbb::global_control' [clang-diagnostic-error]
  tbb::global_control initD(tbb::global_control::active_value(tbb::global_control::max_allowed_parallelism), nThrD);
                      ^
/cvmfs/cms-ib.cern.ch/nweek-02677/slc7_amd64_gcc900/external/tbb/2020_U3-ljfedo/include/tbb/global_control.h:33:5: note: candidate constructor not viable: no known conversion from 'std::size_t' (aka 'unsigned long') to 'tbb::interface9::global_control::parameter' for 1st argument
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22250

ERROR: Build errors found during clang-tidy run.
```
  inline bool operator!=(const zero_allocator<T1, B1> &a, const zero_allocator<T2, B2> &b) {
              ^
Suppressed 646 warnings (645 in non-user code, 1 with check filters).
--
/cvmfs/cms-ib.cern.ch/nweek-02677/slc7_amd64_gcc900/external/tbb/2020_U3-ljfedo/include/tbb/internal/_concurrent_unordered_impl.h:1407:38: error: no matching function for call to object of type 'tbb::interface5::concurrent_unordered_map_traits<std::pair<unsigned int, unsigned int>, std::pair<std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1, 1000000000>>>, bool>, tbb::interface5::internal::hash_compare<std::pair<unsigned int, unsigned int>, edm::StdPairHasher, std::equal_to<std::pair<unsigned int, unsigned int>>>, tbb::tbb_allocator<std::pair<const std::pair<unsigned int, unsigned int>, std::pair<std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1, 1000000000>>>, bool>>>, false>::hash_compare' (aka 'tbb::interface5::internal::hash_compare<std::pair<unsigned int, unsigned int>, edm::StdPairHasher, std::equal_to<std::pair<unsigned int, unsigned int>>>') [clang-diagnostic-error]
        sokey_t hash_key = (sokey_t) my_hash_compare(*pkey);
                                     ^
/cvmfs/cms-ib.cern.ch/nweek-02677/slc7_amd64_gcc900/external/tbb/2020_U3-ljfedo/include/tbb/internal/_concurrent_unordered_impl.h:1080:16: note: in instantiation of function template specialization 'tbb::interface5::internal::concurrent_unordered_base<tbb::interface5::concurrent_unordered_map_traits<std::pair<unsigned int, unsigned int>, std::pair<std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1, 1000000000>>>, bool>, tbb::interface5::internal::hash_compare<std::pair<unsigned int, unsigned int>, edm::StdPairHasher, std::equal_to<std::pair<unsigned int, unsigned int>>>, tbb::tbb_allocator<std::pair<const std::pair<unsigned int, unsigned int>, std::pair<std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1, 1000000000>>>, bool>>>, false>>::internal_insert<tbb::internal::bool_constant<true>, tbb::internal::bool_constant<true>, std::pair<const std::pair<unsigned int, unsigned int>, std::pair<std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long, std::ratio<1, 1000000000>>>, bool>>>' requested here
--
/cvmfs/cms-ib.cern.ch/nweek-02677/slc7_amd64_gcc900/external/tbb/2020_U3-ljfedo/include/tbb/internal/_tbb_hash_compare_impl.h:42:25: error: type 'const edm::StdPairHasher' does not provide a call operator [clang-diagnostic-error]
        return ((size_t)my_hash_object(key));
                        ^
/cvmfs/cms-ib.cern.ch/nweek-02677/slc7_amd64_gcc900/external/tbb/2020_U3-ljfedo/include/tbb/internal/_concurrent_unordered_impl.h:1469:38: note: in instantiation of member function 'tbb::interface5::internal::hash_compare<std::pair<unsigned int, unsigned int>, edm::StdPairHasher, std::equal_to<std::pair<unsigned int, unsigned int>>>::operator()' requested here
--
FWCore/Utilities/interface/std_pair_hasher.h:12:15: error: overloaded 'operator==' must be a binary operator (has 3 parameters) [clang-diagnostic-error]
  inline bool operator==(const zero_allocator<T1, B1> &a, const zero_allocator<T2, B2> &b) {
              ^
FWCore/Utilities/interface/std_pair_hasher.h:16:15: error: overloaded 'operator!=' must be a binary operator (has 3 parameters) [clang-diagnostic-error]
  inline bool operator!=(const zero_allocator<T1, B1> &a, const zero_allocator<T2, B2> &b) {
              ^
Suppressed 802 warnings (801 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22251

- This PR adds an extra 84KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22251/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22251/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

gartung committed:
code-format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22252

- This PR adds an extra 84KB to repository

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

gartung commented:
requires https://github.com/cms-sw/cmsdist/pull/6792

makortel commented on FWCore/Utilities/interface/hash_combine.h:
```diff
@@ -38,6 +38,7 @@ namespace edm {
     hash_combine(seed, args...);
     return seed;
   }
+
```
Could you remove this line?

gartung replied:
ok

makortel commented on FWCore/Utilities/interface/std_pair_hasher.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef FWCore_Utilities_std_pair_hasher_h
+#define FWCore_Utilities_std_pair_hasher_h
```
I'd rename the file `StdPairHasher.h' to be consistent with the struct name.

gartung replied:
OK

makortel commented on FWCore/Utilities/interface/std_pair_hasher.h:
```diff
@@ -0,0 +1,19 @@
+ tbb::hash was changed to used std::hash which does not have an implementation for std::pair.
+ This hasher is taken from the boost::hash implementation.
+*/
+#include "FWCore/Utilities/interface/zero_allocator.h"
```
Shouldn't
```suggestion
#include "FWCore/Utilities/interface/hash_combine.h"
#include <string>
#include <utility>
```
be sufficient?

gartung committed:
Fix header check

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22262

ERROR: Build errors found during clang-tidy run.
```
FWCore/Utilities/interface/StdPairHasher.h:7:10: error: 'FWCore/Utilities/interface/hash_compare.h' file not found [clang-diagnostic-error]
#include "FWCore/Utilities/interface/hash_compare.h"
         ^
Suppressed 461 warnings (460 in non-user code, 1 with check filters).
--
FWCore/Utilities/interface/StdPairHasher.h:7:10: error: 'FWCore/Utilities/interface/hash_compare.h' file not found [clang-diagnostic-error]
#include "FWCore/Utilities/interface/hash_compare.h"
         ^
Suppressed 604 warnings (603 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

gartung commented:
code-checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22263

- This PR adds an extra 140KB to repository

- Found files with invalid states:
  - FWCore/Utilities/interface/std_pair_hasher.h:
    - Added: 6c811a361bcf6c0685a47c8da7d553878f6b60b0
    - Deleted: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/summary.html
**COMMIT**: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4
**CMSSW**: CMSSW_12_0_X_2021-04-23-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33474/14560/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testCalibTrackerSiStripCommon had ERRORS
---> test runtestRecoTauTagRecoTau had ERRORS
---> test materialBudgetTrackerPlots had ERRORS
---> test materialBudgetHGCalPlots had ERRORS
and more ...
</pre>

## AddOn Tests

- [fastsim](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/fastsim)```cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Sat Apr 24 13:31:36 2021-date Sat Apr 24 13:25:40 2021 s - exit: 35584```
- [hlt_data_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_data_Fake2)```cmsDriver.py RelVal -s HLT:Fake2,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_Fake2 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --processName=HLTRECO --filein file:RelVal_Raw_Fake2_DATA.root --fileout file:RelVal_Raw_Fake2_DATA_HLT_RECO.root : FAILED - time: date Sat Apr 24 13:32:28 2021-date Sat Apr 24 13:25:45 2021 s - exit: 35584```
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_data_HIon)```cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run3_data_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT,HLTrigger/Configuration/CustomConfigs.CTPPSRun2Geometry --era Run3_pp_on_PbPb --processName=HLTRECO --filein file:RelVal_Raw_HIon_DATA.root --fileout file:RelVal_Raw_HIon_DATA_HLT_RECO.root : FAILED - time: date Sat Apr 24 13:34:32 2021-date Sat Apr 24 13:25:55 2021 s - exit: 35584```
<details>
<summary>Expand to see more addon errors ...</summary>

- [fastsim1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/fastsim1)
- [fastsim2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/fastsim2)
- [hlt_mc_Fake](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_mc_Fake)
- [hlt_mc_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_mc_Fake2)
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_mc_Fake1)
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_mc_Fake1)
- [hlt_mc_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14560/addOnTests/hlt_mc_PIon)
</details>



## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2877046
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2877017
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 37 files compared)
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-23-2300+a7c094/42391/triggerResults): no differences found

gartung commented:
please test with https://github.com/cms-sw/cmsdist/pull/6792

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14562/summary.html
**COMMIT**: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4
**CMSSW**: CMSSW_12_0_X_2021-04-24-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33474/14562/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - cms-sw/cmsdist#6838 @cms-sw: Update professor2 patch for PPC

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14562/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14562/git-merge-result

Found compilation warnings

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2877486
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2877451
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 37 files compared)
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-24-1100+a7c094/42392/triggerResults): no differences found

makortel commented:
+1

slava77 commented:
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33474	f856889
- code changes are somewhat technical for reco: a switch from task_scheduler_init to global_control
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

silviodonato commented:
kind reminder @cms-sw/db-l2 @cms-sw/hlt-l2

fwyzard commented:
I can check the HLT side.

It looks like there is at least one comment not addressed here, though: https://github.com/cms-sw/cmssw/pull/33474/files#r619203756 ?

fwyzard commented on HLTrigger/Timer/plugins/FastTimerService.cc:
```diff
@@ -767,7 +767,7 @@ void FastTimerService::PlotsPerJob::fill_lumi(AtomicResources const& data, unsig
 ///////////////////////////////////////////////////////////////////////////////
 
 FastTimerService::FastTimerService(const edm::ParameterSet& config, edm::ActivityRegistry& registry)
-    : tbb::task_scheduler_observer(true),
```
this was introduced in #33261; I don't know what the change does, but maybe @dan131riley can comment ?

gartung replied:
I think this was needed to prevent the coroutine from being deleted before this service ended. The early deletion caused a segfault.

gartung replied:
@dan131riley can you confirm.

gartung replied:
See https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/FWCore/Services/plugins/InitRootHandlers.cc#L75

dan131riley replied:
onetbb_2021 removed the constructor that takes a global/local boolean.  While the comments and docs still mention global observers, that actually isn't an option anymore--the only. constructors are a default (that actually does nothing) and one that takes a `task_arena&` to observe a specific arena other than the current one.  See

https://github.com/oneapi-src/oneTBB/blob/2dba2072869a189b9fdab3ffa431d3ea49059a19/include/oneapi/tbb/task_scheduler_observer.h#L74-L85
https://spec.oneapi.com/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_scheduler_observer_cls.html

In fact, we could (and probably should) just remove that line.

gartung replied:
So `public tbb::task_scheduler_observer` should be removed in both locations?

gartung replied:
`ThreadTracker` and `FastTimerService`?

makortel replied:
The member variable must stay, but the explicit call to the default constructor can be removed.

dan131riley replied:
It's actually a parent class, and indeed the parent class needs to stay, it's just the explicit call to the constructor, the `tbb::task_scheduler_observer()`,  that can be removed, since the default constructor will be called anyway.

makortel commented on FWCore/Framework/test/dependentrecord_t.cppunit.cc:
```diff
@@ -38,6 +38,7 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include <cstring>
```
While you're doing changes, could you remove this include as it should not be needed now?

gartung committed:
Remove unneeded #include <cstring>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22357

- This PR adds an extra 144KB to repository

- Found files with invalid states:
  - FWCore/Utilities/interface/std_pair_hasher.h:
    - Added: 6c811a361bcf6c0685a47c8da7d553878f6b60b0
    - Deleted: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

fwyzard commented on HLTrigger/Timer/plugins/FastTimerService.cc:
```diff
@@ -767,7 +767,7 @@ void FastTimerService::PlotsPerJob::fill_lumi(AtomicResources const& data, unsig
 
 FastTimerService::FastTimerService(const edm::ParameterSet& config, edm::ActivityRegistry& registry)
-    : tbb::task_scheduler_observer(true),
+    : tbb::task_scheduler_observer(),
       // configuration
```
so
```suggestion
    :  // configuration
```

gartung replied:
There are two functions that are inherited from task_scheduler_observer according to comment in the file.

fwyzard replied:
the change does not remove the inheritance, only the (unnecessary) explicit call to the default constructor.

gartung committed:
Update HLTrigger/Timer/plugins/FastTimerService.cc

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22358

- This PR adds an extra 24KB to repository

- Found files with invalid states:
  - FWCore/Utilities/interface/std_pair_hasher.h:
    - Added: 6c811a361bcf6c0685a47c8da7d553878f6b60b0
    - Deleted: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/summary.html
**COMMIT**: 1435f32c0226f5de9fc16e8e425628bb52cbf6e3
**CMSSW**: CMSSW_12_0_X_2021-04-28-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33474/14679/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test runtestRecoTauTagRecoTau had ERRORS
---> test materialBudgetTrackerPlots had ERRORS
---> test materialBudgetHGCalPlots had ERRORS
</pre>

## AddOn Tests

- [hlt_data_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_data_Fake2)```cmsDriver.py RelVal -s HLT:Fake2,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_Fake2 --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --processName=HLTRECO --filein file:RelVal_Raw_Fake2_DATA.root --fileout file:RelVal_Raw_Fake2_DATA_HLT_RECO.root : FAILED - time: date Wed Apr 28 20:20:26 2021-date Wed Apr 28 20:14:44 2021 s - exit: 35584```
- [hlt_data_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_data_PRef)```cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run3_hlt_PRef --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run3 --eventcontent RAW --fileout file:RelVal_Raw_PRef_DATA.root --filein /store/data/Run2018D/EphemeralHLTPhysics1/RAW/v1/000/323/775/00000/2E066536-5CF2-B340-A73B-209640F29FF6.root --customise_commands='if hasattr(process,"simMuonGEMPadTask"): setattr(process,"simMuonGEMPadTask",cms.Task())' : FAILED - time: date Wed Apr 28 20:22:13 2021-date Wed Apr 28 20:14:46 2021 s - exit: 35584```
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_data_HIon)```cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run3_hlt_HIon --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run3_pp_on_PbPb --fileout file:RelVal_Raw_HIon_DATA.root --filein /store/data/Run2018D/HIMinimumBias0/RAW/v1/000/325/112/00000/660F62BB-9932-D645-A4A4-0BBBDA3963E8.root --customise_commands='if hasattr(process,"simMuonGEMPadTask"): setattr(process,"simMuonGEMPadTask",cms.Task()); from FWCore.ParameterSet.MassReplace import massSearchReplaceAnyInputTag; massSearchReplaceAnyInputTag(process.SimL1Emulator,"rawDataCollector","rawDataRepacker",False,True)' : FAILED - time: date Wed Apr 28 20:21:55 2021-date Wed Apr 28 20:14:47 2021 s - exit: 35584```
<details>
<summary>Expand to see more addon errors ...</summary>

- [fastsim2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/fastsim2)
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_data_PIon)
- [hlt_mc_Fake](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_mc_Fake)
- [hlt_mc_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_mc_Fake2)
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_mc_Fake1)
- [hlt_data_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14679/addOnTests/hlt_data_Fake1)
</details>



## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2877605
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2877569
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-28-1100+a7c094/42485/triggerResults): no differences found

gartung commented:
please test with https://github.com/cms-sw/cmsdist/pull/6792

gartung commented:
@makortel When I removed the `#include <cstring>` from the Framework test file the clang build error comes up again.
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14687/clang-logs/FWCore/Framework/log.html

gartung committed:
Merge remote-tracking branch 'official-cmssw/master' into gartung-tbb2021

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22380

- This PR adds an extra 212KB to repository

- Found files with invalid states:
  - FWCore/Utilities/interface/std_pair_hasher.h:
    - Added: 6c811a361bcf6c0685a47c8da7d553878f6b60b0
    - Deleted: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @cmsbuild, @slava77, @ggovi can you please check and sign again.

gartung commented:
please test with cms-sw/cmsdist#6792

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14712/summary.html
**COMMIT**: 32448ad3295ff348249347f32900ce60b7076a2e
**CMSSW**: CMSSW_12_0_X_2021-04-29-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33474/14712/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - cms-sw/cmsdist#6855 @hqucms: Enable DYNAMIC_CPU_ARCH in ONNXRuntime to use up to AVX2
 - cms-sw/cmsdist#6854 @gartung: Profiling tools update. 
 - cms-sw/cmsdist#6844 @cms-sw: Update tag for RecoBTag-Combined to V01-11-00

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14712/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14712/git-merge-result


## Build

I found compilation error when building: 

<pre>In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc900/external/cppunit/1.15.x-bcolbf2/include/cppunit/TestCase.h:6,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc900/external/cppunit/1.15.x-bcolbf2/include/cppunit/TestCaller.h:5,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc900/external/cppunit/1.15.x-bcolbf2/include/cppunit/extensions/HelperMacros.h:9,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/dependentrecord_t.cppunit.cc:36:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/dependentrecord_t.cppunit.cc: In member function 'void testdependentrecord::oneOfTwoRecordTest()':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/dependentrecord_t.cppunit.cc:1337:27: error: 'strstr' was not declared in this scope
 1337 |       CPPUNIT_ASSERT(0 != strstr(e.what(), "DepOn2Record"));
      |                           ^~~~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc900/external/cppunit/1.15.x-bcolbf2/include/cppunit/TestAssert.h:274:37: note: in definition of macro 'CPPUNIT_ASSERT'
  274 |   ( CPPUNIT_NS::Asserter::failIf( !(condition),                                   \
      |                                     ^~~~~~~~~
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/interval_t.cppunit.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/proxyfactoryproducer_t.cppunit.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/iovsyncvalue_t.cppunit.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/intersectingiovrecordintervalfinder_t.cppunit.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/eventsetupscontroller_t.cppunit.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/dependentrecord_t.cppunit.cc:1337:27: error: use of undeclared identifier 'strstr'
      CPPUNIT_ASSERT(0 != strstr(e.what(), "DepOn2Record"));
                          ^
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/FWCore/Framework/test/dependentrecord_t.cppunit.cc:1338:27: error: use of undeclared identifier 'strstr'
      CPPUNIT_ASSERT(0 != strstr(e.what(), "Dummy2Record"));
                          ^
</pre>

makortel commented:
> When I removed the `#include <cstring>` from the Framework test file the clang build error comes up again.
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14687/clang-logs/FWCore/Framework/log.html

My bad, could you add it back then?

gartung committed:
Revert "Remove unneeded #include <cstring>"

This reverts commit a09f99ba36b7c52cbc33587e06243e4087e1f787.

gartung commented:
please test with cms-sw/cmsdist#6792

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33474/22385

- This PR adds an extra 168KB to repository

- Found files with invalid states:
  - FWCore/Utilities/interface/std_pair_hasher.h:
    - Added: 6c811a361bcf6c0685a47c8da7d553878f6b60b0
    - Deleted: f85688952ad47e0a25f06cb7cf5a85cb5cce31a4

cmsbuild commented:
Pull request #33474 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @fwyzard, @ianna, @mdhildreth, @Martin-Grunewald, @jpata, @slava77, @ggovi can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14724/summary.html
**COMMIT**: 3132e55be209f22bbea546afe4d48f11df2c8e73
**CMSSW**: CMSSW_12_0_X_2021-04-29-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33474/14724/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - cms-sw/cmsdist#6855 @hqucms: Enable DYNAMIC_CPU_ARCH in ONNXRuntime to use up to AVX2
 - cms-sw/cmsdist#6854 @gartung: Profiling tools update. 
 - cms-sw/cmsdist#6849 @cms-sw: Update SCRAM V3
 - cms-sw/cmsdist#6844 @cms-sw: Update tag for RecoBTag-Combined to V01-11-00

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14724/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14724/git-merge-result


## Build

I found compilation error when building: 

<pre>/cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-29-1100/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so: undefined reference to `tbb::task_group_context::~task_group_context()'
/cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-29-1100/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so: undefined reference to `tbb::internal::internal_resume(void*)'
/cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-29-1100/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so: undefined reference to `tbb::assertion_failure(char const*, int, char const*, char const*)'
/cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-29-1100/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so: undefined reference to `tbb::internal::allocate_root_with_context_proxy::allocate(unsigned long) const'
/cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-29-1100/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so: undefined reference to `tbb::interface7::internal::task_arena_base::internal_max_concurrency(tbb::interface7::task_arena const*)'
collect2: error: ld returned 1 exit status
>> Deleted: tmp/slc7_amd64_gcc900/src/DataFormats/Provenance/test/ElementID_t/ElementID_t
gmake: *** [tmp/slc7_amd64_gcc900/src/DataFormats/Provenance/test/ElementID_t/ElementID_t] Error 1
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_0_X_2021-04-29-1100/src/DataFormats/Provenance/test/EntryDescription_t.cpp
>> Building binary EntryDescription_t
/cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: warning: libtbb_preview.so.2, needed by /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-29-1100/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so, not found (try using -rpath or -rpath-link)
</pre>

gartung commented:
test parameters:
full=true

gartung commented:
please test with cms-sw/cmsdist#6792

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14760/summary.html
**COMMIT**: 3132e55be209f22bbea546afe4d48f11df2c8e73
**CMSSW**: CMSSW_12_0_X_2021-04-30-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33474/14760/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - cms-sw/cmsdist#6860 @cms-sw: [SCRAMV3] Fix for scram tool info

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14760/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a7c094/14760/git-merge-result

Found compilation warnings

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2662646
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2662611
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-30-1100+a7c094/42552/triggerResults): no differences found

cvuosalo commented:
+1

makortel commented:
+1

fwyzard commented:
+1

perrotta commented:
+1
- Technical, already approved by Core
- Jenkins tests pass

silviodonato commented:
kind reminder @cms-sw/db-l2

ggovi commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1

silviodonato commented:
See https://github.com/cms-sw/cmssw/issues/33642 . 
Actually I noticed the warnings, but I though it was not related to this PR since the PR does not change anything in `EventFilter/Utilities`
